File: app-rn/src/storage/savedState.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { useCallback, useEffect, useRef, useState } from 'react';
import { content } from '../generated/content';
import type { DayKey, SavedState, WordProgress } from '../core';

const KEY = 'tamil-train/state/v1';

export function defaultState(): SavedState {
  const t = content.rules.trip;
  return {
    schema: 1,
    settings: { length: t.defaultLength, choices: t.defaultChoices, actionStops: t.actionStopsDefaultOn, showGloss: t.showEnglishGlossDefault },
    progress: {},
    tripsByDay: {},
  };
}

/**
 * Whole app state in one JSON blob. Small (a few KB) so read/write-all is fine.
 * The iOS app stores the identical JSON (ProgressStore.swift), so a grown-up
 * can export from one device and import on another.
 */
export function useSavedState() {
  const [state, setState] = useState<SavedState | null>(null);
  const latest = useRef<SavedState | null>(null);

  useEffect(() => {
    AsyncStorage.getItem(KEY)
      .then((raw) => {
        const loaded = raw ? { ...defaultState(), ...(JSON.parse(raw) as SavedState) } : defaultState();
        latest.current = loaded;
        setState(loaded);
      })
      .catch(() => { latest.current = defaultState(); setState(latest.current); });
  }, []);

  const update = useCallback((fn: (s: SavedState) => SavedState) => {
    if (!latest.current) return;
    latest.current = fn(latest.current);
    setState(latest.current);
    AsyncStorage.setItem(KEY, JSON.stringify(latest.current)).catch(() => {});
  }, []);

  // Synchronous read for the trip engine (it scores inside event handlers).
  const getProgress = useCallback(() => latest.current?.progress ?? {}, []);
  const saveProgress = useCallback(
    (id: string, p: WordProgress) => update((s) => ({ ...s, progress: { ...s.progress, [id]: p } })),
    [update],
  );
  const countTrip = useCallback(
    (day: DayKey) => update((s) => ({ ...s, tripsByDay: { ...s.tripsByDay, [day]: (s.tripsByDay[day] ?? 0) + 1 } })),
    [update],
  );

  return { state, update, getProgress, saveProgress, countTrip };
}

File: ios/TamilTrainCore/Sources/TamilTrainCore/SavedState.swift
import Foundation

/// Persisted shape. Byte-compatible with the React Native app's AsyncStorage
/// JSON (app-rn/src/storage/savedState.ts), so progress can move between devices.
public struct SavedState: Codable, Equatable, Sendable {
    public var schema: Int
    public var settings: TripSettings
    public var progress: [String: WordProgress]
    public var tripsByDay: [DayKey: Int]

    public static func defaults(_ rules: Rules) -> SavedState {
        SavedState(
            schema: 1,
            settings: TripSettings(length: rules.trip.defaultLength, choices: rules.trip.defaultChoices,
                                   actionStops: rules.trip.actionStopsDefaultOn, showGloss: rules.trip.showEnglishGlossDefault),
            progress: [:],
            tripsByDay: [:]
        )
    }

    public func status(of word: Word, rules: ScoringRules) -> WordStatus {
        Tracer.status(progress[word.id], rules: rules)
    }

    public func knownCount(_ words: [Word], rules: ScoringRules) -> Int {
        words.filter { status(of: $0, rules: rules) == .known }.count
    }
}

/// One small JSON file in Application Support. Atomic writes; a few KB total.
public final class ProgressStore {
    public let url: URL

    public init(fileName: String = "tamil-train-state-v1.json") {
        let dir = (try? FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true))
            ?? FileManager.default.temporaryDirectory
        self.url = dir.appendingPathComponent(fileName)
    }

    public init(url: URL) {
        self.url = url
    }

    public func load() -> SavedState? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(SavedState.self, from: data)
    }

    public func save(_ state: SavedState) {
        let enc = JSONEncoder()
        enc.outputFormatting = [.sortedKeys]
        guard let data = try? enc.encode(state) else { return }
        try? data.write(to: url, options: .atomic)
    }
}
